kmod/core/core.c

```diff
@@ -376,17 +408,29 @@ static int kpatch_remove_patch(void *data)
 	if (ret == KPATCH_STATE_FAILURE)
 		return -EBUSY;
 
+	/* run any user-defined pre-unpatch callbacks */
+	list_for_each_entry(object, &kpmod->objects, list) {
+		if (!kpatch_object_linked(object) ||
+		    !object->pre_unpatch_callback ||
+		    !object->callbacks_enabled)
+			continue;
+
+		(*object->pre_unpatch_callback)(object);
+	}
```

Comment:
This needs to be done *before* the state is set to KPATCH_STATE_SUCCESS, which is when the patch removal goes live to NMIs.

Reply:
OK - I wasn't too clear about the NMI interaction, but that makes sense.

Reply:
BTW, what then happens if we run pre-unpatch callbacks before the NMI check and then the latter fails?  We're still patched, but already ran those callbacks.

Reply:
I guess it's similar to the case in livepatch where the pre-unpatch callback is called, but the unpatching never completes (or is reversed).  Since the post-patch callback is the reverse of the pre-unpatch callback, should we call the post-patch callback if the NMI check fails?

Reply:
I think so.  Just for the record, this is getting weird.